### ISSUE TITLE
Not possible to disable dynamic-import

### ISSUE THREAD

**Issue by mitar** [meteor/meteor#10704]
It looks like on Meteor 1.8.1 it is not possible to disable `dynamic-import` because it is a strict dependency of `ecmascript`. This is especially problematic because `dynamic-import` forces CSP to allow `eval`. So if you want to disable `eval` you have to remove `ecmascript`?

Moreover, if one does not want to use `dynamic-import` (because it does not work with disabled `eval`), having it bundled for the client always seems unnecessary.

See here for more information: https://github.com/meteor/meteor/blob/devel/packages/dynamic-import/security.js

**Comment by nathan-muir** [comment]
Yeah, would love to get rid of `dynamic-import` or have it create static files that can be CDN hosted.

https://github.com/meteor/meteor-feature-requests/issues/296

**Comment by mitar** [comment]
Does `WebAppInternals.setBundledJsCssUrlRewriteHook` not work for dynamic import files? For CDN hosting?

**Comment by nathan-muir** [comment]
AFAIK `dynamic-import` does a HTTP POST (used to be DDP) to the meteor server, which then generates a mini-bundle of the requested paths on the fly.

**Comment by jamesmillerburgess** [comment]
On Galaxy we use `dynamic-import` with a CDN by setting that hook.

**Comment by jamesmillerburgess** [comment]
Sorry, @nathan-muir is right - these calls don't go over the CDN, but it would be great if they could.

**Comment by mitar** [comment]
> which then generates a mini-bundle of the requested paths on the fly.

So then all possible combinations of files should be pre-generated so that they can be stored statically on CDN? Such generation could be useful already for local serving as well, instead of dynamic generation.

Anyway, I think dynamic-import support for CDN should probably be another issue.

**Comment by coagmano** [comment]
Just noticed that `socket-stream-client` also has a strict dependency on `dynamic-import`, which means any app that uses `ddp` also gets it 😱

Interestingly, it seems to be redundant at the moment as it's used to [import `SockJS` as a fallback here](https://github.com/meteor/meteor/blob/5c32331591f268f2acdac5a1a7481eb9e65dc3d4/packages/socket-stream-client/browser.js#L198), but is already [statically imported here](https://github.com/meteor/meteor/blob/5c32331591f268f2acdac5a1a7481eb9e65dc3d4/packages/socket-stream-client/browser.js#L11)...

**Comment by coagmano** [comment]
Actually, is there any reason why `modules` needs to imply `dynamic-import` except convenience? 
Can that one line just be removed without consequence?

**Comment by mitar** [comment]
Still relevant.

**Comment by sebakerckhof** [comment]
@mitar regarding the eval-problem, you can add a file to your server overwriting the dynamic-import:
```js
import { Meteor } from 'meteor/meteor';
Meteor.startup(() => {
  BrowserPolicy.content.disallowEval();
});
```

If you don't have dynamic imports, this should not break anything.

**Comment by FishSaidNo** [comment]
@sebakerckhof Thanks for that. Seems `BrowserPolicy.content.disallowEval()` overwrite only applies when called within a Meteor.startup block, which isn't mentioned in the browser-policy docs.

**Comment by filipenevola** [comment]
I'm closing this because Seba provided a workaround if the workaround is not enough for you then please open a new issue referencing this one.

**Comment by mitar** [comment]
Sorry, what? The issue is how to disable `dynamic-import`. Where is a workaround for this?

**Comment by mitar** [comment]
So the issue is that I want to get rid of `dynamic-import`. It is hard to get rid of that and make sure no dependency uses it, without removing `dynamic-import` package. If one can get rid of `dynamic-import` I can be sure it is not being used. If `dynamic-import` is around, then I have to manually inspect all my dependencies to see if they are using `dynamic-import` or not, for real. Also, just disabling `eval` means that it can happen that some dependency uses `dynamic-import` and then it starts failing.

Really not good experience.

I suggest `dynamic-import` is not forced on us when one uses `ecmascript`.

**Comment by filipenevola** [comment]
I was assuming the workaround was enough because of this:
> This is especially problematic because dynamic-import forces CSP to allow eval. So if you want to disable eval you have to remove ecmascript?

But I can reopen it, no problem 😉

**Comment by mitar** [comment]
I mean, this is important motivation. But sadly just disabling eval without removing `dynamic-import` introduces potential issues as described above because you do not know which of your dependencies might use it.

Thanks for reopening.

**Comment by mitar** [comment]
Still relevant.

**Comment by StorytellerCZ** [comment]
We will have to make sure that this is taken care of with upgrading browser policy package.
Discussion: https://github.com/meteor/meteor/discussions/11591

**Comment by paulincai** [comment]
End of 2023 this is still very much relevant. Dynamic imports contributes 10% to my chat app with no way of removing it.
I think it makes sense to consider that unde a 1.5MB uncompressed most of us wouldn't consider to split code.

![Screenshot 2023-11-13 at 8 23 21 PM](https://github.com/meteor/meteor/assets/26188915/2395beb5-fdcf-42e7-8e87-64a0e6d6d0dd)

**Comment by jankapunkt** [comment]
Coming back to this, I analyzed the whole meteor repo and this is what I found:

- dynamic import is hard-wired with `ecmascript` and `typescript`
- it's also hard-wired by finish upgraders and will always be added by `meteor update`
- making it a weak dependency works but it needs to be removed in secondary packages like Blaze as well (I think I can make a PR for this)
- bundle-visualizer is also hard-wired using `import()` to load all requred data after browser bundle is resolved and started, I think in order to not include styles and code from itself to the chart
- I can try to unwire it from bundle-visualizer
- some skeletons and files in the tool have it linked
- I can open a Draft PR and we can find a solution that does not break everyones app after upgrade but I'm skeptical, it's very deep linked with the system
- @Grubba27 @denihs @henriquealbert this is a perfect example on tight coupling that should be avoided in future design decisions
- also pülease forgive me, I know this sounds talking from high ground and I admit, I would have done a much worse implementation: it's simply my analysis of the situation

**Comment by jankapunkt** [comment]
Fix attempt is here: https://github.com/meteor/meteor/pull/12879
